openshift/s3-backup/docker/prune.py

```diff
@@ -45,51 +45,53 @@ def __init__(self, desired_backups: int, interval: timedelta) -> None:
         interval: Desired interval (in days) between backups
         """
         self.desired_backups = desired_backups
-        self.backups_found = 0
-        self.prev_timestamp = None
         self.interval = interval
-
-    def _is_keeper(self, timestamp: datetime) -> bool:
-        """ Run through all the conditions to decide if we keep this timestamp or not. """
-        if self.backups_found < self.desired_backups:
-            # If we don't have any backups yet, keep it!
-            if self.prev_timestamp is None:
-                return True
-            # If this backup is older than the previous interval, keep it!
-            if self.prev_timestamp - timestamp >= self.interval:
-                return True
-        return False
-
-    def is_keeper(self, timestamp: datetime) -> bool:
-        """ Decide if we should keep this timestamp or not. """
-        if self._is_keeper(timestamp):
-            self.prev_timestamp = timestamp
-            self.backups_found += 1
-            return True
-        return False
+        self.files_to_keep = []
+
+    def evaluate(self, new_filename) -> None:
+        """ Consider the file, and manage the list of files we're deciding to keep. """
+        self.files_to_keep.append(new_filename)
+        self.files_to_keep.sort()
+        # If we have more files than we desire:
+        if len(self.files_to_keep) > self.desired_backups:
+            prev_timestamp = None
+            for file in self.files_to_keep:
+                timestamp = extract_datetime(file)
+                if prev_timestamp is None:
+                    prev_timestamp = timestamp
+                    continue
+                # If the time difference between two files is less than the desired interval:
+                if timestamp - prev_timestamp < self.interval:
+                    self.files_to_keep.remove(file)
+                    break
+                # If the time difference between two files is greater than or equal to the interval:
+                if timestamp - prev_timestamp >= self.interval:
+                    prev_timestamp = timestamp
+                    continue
+            # If we have too many files, get rid of the oldest one:
+            if len(self.files_to_keep) > self.desired_backups:
+                self.files_to_keep.pop(0)
```

Comment:
if for some reason this pruning script stopped running for a while, this wouldn't clean up all of the oldest files.

E.g., say `len(self.files_to_keep) = 9` and `self.desired_backups = 5` for hourly backups, then this logic would only remove 1 or 2 files (depending on their timestamps), so we'd still end up with `len(self.files_to_keep)` = 7 or 8.

So, critical question - how likely is it that this script would fail to run (or be paused)?

Reply:
The logic is slightly different, interruptions don't affect the result because the logic you're referring to doesn't start with the full list. It kinda goes like this:

- 1 get full list from object store
- 2 take file 0, apply logic. take file 1, apply logic ... take file n, apply logic
- 
So the `desired_backups` variable always starts empty, and then starts adding files one by one and considering them. It should never exceed 5 because of that check.

Regardless, I've added a test case to ensure that what you describe doesn't happen.